skops/card/_card_alternative.py

```diff
@@ -0,0 +1,361 @@
+from __future__ import annotations
+
+import re
+from dataclasses import dataclass, field
+from pathlib import Path
+from reprlib import Repr
+from typing import Any, Iterator, Protocol
+
+from huggingface_hub import CardData
+from sklearn.utils import estimator_html_repr
+from tabulate import tabulate  # type: ignore
+
+from skops.card._model_card import PlotSection, TableSection
+
+aRepr = Repr()
+aRepr.maxother = 79
+aRepr.maxstring = 79
+
+
+def split_subsection_names(key: str) -> list[str]:
+    return key.split("/")
+
+
+def _clean_table(table: str) -> str:
+    # replace line breaks "\n" with html tag <br />, however, leave end-of-line
+    # line breaks (eol_lb) intact
+    eol_lb = "|\n"
+    placeholder = "$%!?"  # arbitrary sting that never appears naturally
+    table = (
+        table.replace(eol_lb, placeholder)
+        .replace("\n", "<br />")
+        .replace(placeholder, eol_lb)
+    )
+    return table
+
+
+@dataclass
+class Section:
+    title: str
+    content: Formattable | str | None = None
+    subsections: dict[str, Section] = field(default_factory=dict)
+
+
+class Formattable(Protocol):
+    def format(self) -> str:
+        ...
+
+
+class Card:
+    @classmethod
+    def make_default(
+        cls, model, model_diagram: bool = True, metadata: CardData | None = None
+    ):
+        """Add a bunch of default sections, yet to be implemented"""
+        raise NotImplementedError
+
+    def __init__(
+        self, model, model_diagram: bool = True, metadata: CardData | None = None
+    ):
+        self.model = model
+        self.model_diagram = model_diagram
+        self.metadata = metadata or CardData()
+
+        self._data: dict[str, Section] = {}
```

Comment:
One thought: Might be worth using an `OrderedDict` for this (and for the `subsections` field in the `Section` class.)

That way it would allow us to have more control over the ordering of the output. I know that as of 3.7, the order of the keys in a `dict` is just the order of insertion, but it might be nice to offer users the ability to insert between existing sections.

Reply:
Do we need to support that? I see it as "the card is rendered in the order you're adding things" kinda API, especially since we don't expose this `_data` to the users as a public API.

Reply:
Yes, one weakness of the chosen data structure is that it's hard to change the order of sections once they are created. However, I'm not sure how much `OrderedDict` would help. I haven't used it in quite a while, but from the docs, the only additional methods it seems to have are `move_to_end` and `popitem(last=...)`, both of which don't help that much? Ideally, we would have something like:

```python
card = Card(...)
card.add(section_a="content a", section_c="content c")
card.insert(section_b="content b", after="section_a")
```

or something. Not sure if that's easier to implement with `OrderedDict` than with normal dicts.

Alternatively, we could implement the underlying tree as a nested list, which gives us more flexibility to modify the order:

```
>>> card._data
[
  ("section_a", "content a"),
  ("section_b", [
    "content b",
    ("section_b0", "content_b0"),
    ("section_b1", "content_b1"),
  ],
  ...,
]
```

The disadvantage is that selecting items will be more difficult an inefficient.

Reply:
> Do we need to support that? I see it as "the card is rendered in the order you're adding things" kinda API, especially since we don't expose this _data to the users as a public API.

I think it would be fine to not support changing ordering for an MVP, but it would be at least a nice-to-have down the road imo. Especially if the default `Card` class users handle has a fixed order for the defaults. Users might want to insert things between other sections.

Reply:
> The disadvantage is that selecting items will be more difficult an inefficient.

I may have been remembering the abilities of OrderedDicts with rose tinted glasses... 
Another alternative that might be a little simpler would be to hold a `subsection_order` attribute in each `Section` that is just a list with the names of children, and renders them in that order.

So if we rendered, as above:
```
("section_b", [
    "content b",
    ("section_b0", "content_b0"),
    ("section_b1", "content_b1"),
  ],
)
```
Section b would just have an attribute like `[__content__, section_b0, section_b1]`

We could implement ease-of-use things like support for `.add(..., after="foo")` as above as well

On efficiency, while it's always good to think about, I don't think we need to get too deep into optimizations with something like this. I'd be very surprised if anyone had even 30 sections total, and at that kind of scale it's not really a limitation imo.

Reply:
> The disadvantage is that selecting items will be more difficult an inefficient.

We can always implement `__getitem__`.

Reply:
> Another alternative that might be a little simpler would be to hold a subsection_order attribute in each Section that is just a list with the names of children, and renders them in that order.

This would be a kind of index. I think if possible, we should avoid it so that we don't leave room for possible bugs where the content and the "index" are out of sync.

> On efficiency, while it's always good to think about, I don't think we need to get too deep into optimizations with something like this.

Probably you're right, the main concern should be usability and simple code. But you never know how users can surprise you. In skorch, we ran into an issue because a user was training a neural net for tens of thousands of epochs, which we didn't expect...

> We can always implement `__getitem__`.

Yes, I thought about that, it just feels "wrong" to have that method when the access is O(n) and not O(1).

Reply:
I feel like nested list is a bit overkill (and selection should be less efficient as you said). I'm in favor of implementation as is with a regular dict.